### DISCUSSION TITLE
New item in a Map set as an attribute in session are not updated in remote cached session with version 14.0.7

### DISCUSSION THREAD

**Discussion by davidmestr** [infinispan/infinispan#10736] [Q&A] [answered]
In  version 14.0.7, the way of save `InfinispanSession` (in _org.infinispan.spring.common.session.AbstractInfinispanSessionRepository_) has changed. So, there is a new mechanism that detect the changes in session and only update them. 

In older versions I can add a new item to an existent Map in session and it is saved in remote cache and can be recovered in next request. The problem is that with new version, the item is only cached if the map is explicitly set as attribute.

As example:
```java
  protected void doGet(final HttpServletRequest request, final HttpServletResponse response) {
    final Map<String, String> map;
    HttpSession ses = request.getSession(false);
    if (ses == null) {
      ses = request.getSession(true);
      map = new HashMap<>();
      ses.setAttribute("MAP", map);
    } else {
      map = (Map<String, String>) ses.getAttribute("MAP");
      if (map.get("new item") == null) {
        LOG.info("create item");
        map.put("new item", "new value");
      }
    }
  }
}
```
The session is created in the first call.  In the following calls the "new item" is always created because it is never found in session.

Are there some configuration or workaround to keep the behaviour of previous versions?

**Comment by tristantarrant** [comment]
Maybe I should bring in @thelateperseus who made the latest modifications to the Spring Session code.

**Comment by thelateperseus** [comment, marked as answer]
@davidmestr you can call `setSaveMode(SaveMode.ALWAYS)` on `AbstractInfinispanSessionRepository` to ensure that the session is always saved at the end of the request processing. From the Javadoc in `SaveMode`:

> Always save all session attributes, regardless of the interaction with the session. In highly concurrent environments, this mode increases the risk of attributes being overwritten during processing of parallel requests.

**Reply by davidmestr** [reply]
Thank you. I can see that may be `SaveMode.ON_GET_ATTRIBUTE` works for us too
